File: Partition Labels.js
var partitionLabels = function(s) {
    var last = {};
    var result = [];
    var end = 0;
    var start = 0;
    for(var i = 0; i < s.length; i++){
        last[s[i]] = i;
    }
    
    for(var i = 0; i < s.length; i++){
        end = Math.max(end, last[s[i]);
        
        if(end === i){
            result.push(end - start + 1);
            start = end +1;
        }
    }
    return result;
};

File: Partition Labels.py
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        result = []
        last = 0
        prev_part_end = -1
        last_occurence = {}
        
        for i, char in enumerate(s):
            last_occurence[char] = i

        for i in range(len(s)):
            last_idx = last_occurence[s[i]]
            last = max(last_idx, last)

            if i == last:
                result.append(last - prev_part_end)
                prev_part_end = last

        return result
